### PR TITLE
Separate alpha and rbg interpolation then recombine to fix issue11316

### PR DESCRIPTION
**Pull request by dabana**
## PR Summary

Addresses issue #11316.

@tacaswell suggested two solutions:

1) to change the *output* array starting as (1, 1, 1, 0)
2) re-sample the RGB and A arrays separatly

I tested both solutions. Solution 1 did not work. I tested various *output* arrays but it does not have any impact on the result: grey boundaries are still visible. Finally, solution was the right one. I interpolate the alpha channel and the 3 rgb channels a 2 separate images. I then recombine into a 4-channel RGBA output.

This is a crude fix to the issue. Help wanted to improve elegance and/or efficiency.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Thanks @dabana Looks promising - I'm surprised this doesn't hit any tests.  Since it doesn't, can you make some?

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -486,16 +486,30 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
 
+                #resample rgb channels
+                A = _rgb_to_rgba(A[...,:3])
                 _image.resample(
                     A, output, t, _interpd_[self.get_interpolation()],
                     self.get_resample(), alpha,
                     self.get_filternorm(), self.get_filterrad())
 
+                #recombine rgb and alpha channels
+                output = np.dstack((output[...,:3], output_a[...,0]))
```
Comment (tacaswell) [inline lib/matplotlib/image.py]: Can we update `output` in place here?  Memory allocation in this code path is a mid-level concern.

File: lib/matplotlib/image.py [hunk 2]
```diff
@@ -486,16 +486,30 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
 
                 alpha = self.get_alpha()
                 if alpha is None:
-                    alpha = 1.0
+                    alpha = 1
+
+                #resample alpha channel
+                alpha_tuple = (A[...,3], A[...,3], A[...,3])
+                alpha_channel = _rgb_to_rgba(np.dstack(alpha_tuple))
+                _image.resample(
```
Comment (tacaswell) [inline lib/matplotlib/image.py]: This will also re sample 2D (1-channel) arrays so we can do this without making a a 4 channel copy.

**Comment by tacaswell** [comment]
Thanks @dabana !  Can you also add a test of this?  The example in original issue would be a perfect test image.

File: lib/matplotlib/tests/test_image.py [hunk 3]
```diff
@@ -42,6 +42,19 @@ def test_image_interps():
     ax3.imshow(X, interpolation='bicubic')
     ax3.set_ylabel('bicubic')
 
+@image_comparison(baseline_images=['interp_alpha'],
+                  extensions=['png'])
```
Comment (jklymak) [inline lib/matplotlib/tests/test_image.py]: Looks great.  Please add `remove_text=True` to the image comparison args to reduce our dependence on fonts...

**Comment by dabana** [comment]
@tacaswell I believe this is ready to merge.

**Comment by timhoffm** [comment]
I don‘t have a complete overview of the topic, but can we save the alpha interpolation if there is no alpha involved?

File: lib/matplotlib/image.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/image.py]: This is still doing the alpha re-sampling 4 times instead of once.

Comment (tacaswell) [reply lib/matplotlib/image.py]: sorry, was reading through commit-by commit and thought I was looking at the total diff, ignore my (now deleted) comment if you got an email.

**Comment by tacaswell** [comment]
@dabana Sorry this sat for so long.  Could you possible resbase / squash this a bit so that we only have one copy of the images you added?

**Comment by dabana** [comment]
> @dabana Sorry this sat for so long. Could you possible resbase / squash this a bit so that we only have one copy of the images you added?

Sorry @tacaswell I don't think I am getting what you mean. I just synced my fork with the base repo and I rebased this branch with the master. There were no conflicts and I do not see any duplicates of my images in the baseline_image directory.

**Comment by jklymak** [comment]
@dabana the 8 commits have different versions of the image files in them.  If we merge as is, each version of the images will be put in the repo separately, despite the fact the last version is the one we want.

**Comment by jklymak** [comment]
For how to squash, the somewhat poorly named link here should help:   Please do the backup step :wink:! https://matplotlib.org/devel/gitwash/development_workflow.html#rewriting-commit-history

**Comment by tacaswell** [comment]
@dabana you missed the `force push` step (the git command line helpfully provides you safe but wrong advise)


I think you want to do

```
git checkout fix-for-issue11316
git resest --hard afa6123  # THIS ONE IS A BIT DANGEROUS, but `git reflog` can save you
git push --force-with-lease   # THIS ONE IS THE DANGEROUS ONE, but more force pushes can help
```

**Comment by tacaswell** [comment]
the docs CI is failing because it runs on the branch head instead of simulating the merge and then running the tests so you don't have the fixes from master branch for these docs build issues.

**Comment by dabana** [comment]
> the docs CI is failing because it runs on the branch head instead of simulating the merge and then running the tests so you don't have the fixes from master branch for these docs build issues.

Is there anything I can do about it to help the merge?

**Comment by tacaswell** [comment]
If you re-base the branch on to current master (and then force-push again) the docs will pass, however I don't think we are going to hold the PR over that.  

We are just waiting for a second core-dev to review and merge this.

**Pull request merged**